core/services/job/models.go

```diff
@@ -848,47 +849,50 @@ type LiquidityBalancerSpec struct {
 }
 
 type WorkflowSpec struct {
-	ID int32 `toml:"-"`
-	// TODO it may be possible to compute the workflow id from the hash(yaml, owner, name) and remove this field
-	WorkflowID    string    `toml:"workflowId"` // globally unique identifier for the workflow, specified by the user
-	Workflow      string    `toml:"workflow"`
-	WorkflowOwner string    `toml:"workflowOwner"` // hex string representation of 20 bytes
-	WorkflowName  string    `toml:"workflowName"`  // 10 byte plain text name
+	ID       int32  `toml:"-"`
+	Workflow string `toml:"workflow"` // the yaml representation of the workflow
+	// fields derived from the yaml spec, used for indexing the database
+	// note: i tried to make these private, but translating them to the database seems to require them to be public
+	WorkflowID    string    `toml:"-" db:"workflow_id"`    // Derived. Do not modify. the CID of the workflow.
+	WorkflowOwner string    `toml:"-" db:"workflow_owner"` // Derived. Do not modify. the owner of the workflow.
+	WorkflowName  string    `toml:"-" db:"workflow_name"`  // Derived. Do not modify. the name of the workflow.
 	CreatedAt     time.Time `toml:"-"`
 	UpdatedAt     time.Time `toml:"-"`
 }
 
 var (
-	ErrInvalidWorkflowID    = errors.New("invalid workflow id")
-	ErrInvalidWorkflowOwner = errors.New("invalid workflow owner")
-	ErrInvalidWorkflowName  = errors.New("invalid workflow name")
+	ErrInvalidWorkflowID = errors.New("invalid workflow id")
 )
 
 const (
 	workflowIDLen = 64 // conveniently the same length as a sha256 hash
-	// owner and name are constrained the onchain representation in [github.com/smartcontractkit/chainlink-common/blob/main/pkg/capabilities/consensus/ocr3/types/Metadata]
-	workflowOwnerLen = 40 // hex string representation of 20 bytes
-	workflowNameLen  = 10 // plain text name
 )
 
-// Validate checks the length of the workflow id, owner and name
-// that latter two are constrained by the onchain representation in [github.com/smartcontractkit/chainlink-common/blob/main/pkg/capabilities/consensus/ocr3/types/Metadata]
-func (w *WorkflowSpec) Validate() error {
-	if len(w.WorkflowID) != workflowIDLen {
-		return fmt.Errorf("%w: incorrect length for id %s: expected %d, got %d", ErrInvalidWorkflowID, w.WorkflowID, workflowIDLen, len(w.WorkflowID))
-	}
+// init initializes the workflow spec by parsing the yaml and setting the id, owner and name
+func (w *WorkflowSpec) init() error {
+	f := sync.OnceValue(func() error {
+		s, err := pkgworkflows.ParseWorkflowSpecYaml(w.Workflow)
+		if err != nil {
+			return fmt.Errorf("failed to parse workflow spec %s: %w", w.Workflow, err)
+		}
+		w.WorkflowOwner = strings.TrimPrefix(s.Owner, "0x") // the json schema validation ensures it is a hex string with 0x prefix, but the database does not store the prefix
+		w.WorkflowName = s.Name
+		w.WorkflowID = s.CID
 
-	w.WorkflowOwner = strings.TrimPrefix(w.WorkflowOwner, "0x")
-	_, err := hex.DecodeString(w.WorkflowOwner)
+		return nil
+	})
+	return f()
```

Comment:
@krehermann This is pretty confusing, we only call f() once, so I'd expect the OnceValue to be doing nothing here. Am I missing something?

Reply:
i'm trying to initialize the derived values from the yaml via some lazy initialization.

i was thinking this ought to be done only once.

the only entrypoint is Validate. and there is nothing preventing Validate from being called many times.

however, as i read, there nothing preventing someone from modifying the value of `Workflow` either, in case it's simply wrong for the init to be done once.


i'll move into validate and get rid of the once.

thanks for catching it